Community Activities: Added basic runtime telemetry for Credentials, Cryptography, Database, FTP and Java activities

coramiam opened the pull request:
Added basic runtime telemetry (execution event only) to the following community activity packages:
- UiPath.Database.Activities
- UiPath.FTP.Activities
- UiPath.Cryptography.Activities
- UiPath.Java.Activities
- UiPath.Credentials.Activities

This PR only covers the infrastructure for Telemetry which was migrated from System.Activities, and a basic runtime event for each activity mentioned in the Jira tickets.

⚠️ To do: 
- in a separate PR - add field details for each activity to be sent to Telemetry.
- later edit: Not a bug, but an activity misconfiguration -> - possible bug - investigate why the container activities (child activities) don't seem to trigger events - for example, the FTP child activities don't seem to trigger events, only the FTP scope activity does. 

**Jira tickets:** 
https://uipath.atlassian.net/browse/STUD-77058
https://uipath.atlassian.net/browse/STUD-77060
https://uipath.atlassian.net/browse/STUD-77061
https://uipath.atlassian.net/browse/STUD-77063
https://uipath.atlassian.net/browse/STUD-77065

**How to test:**
Add the activities in a Studio project, run/debug them and check the Telemetry events sent in the Robot AppInsights page using this filter:

```
requests
| where name contains ".Activities."
| where tostring(customDimensions.ActivityPackage) contains "Database." // change this to your package name
| order by name asc
```

coramiam committed:
Community Activities:

Added basic runtime telemetry (execution event only) to the following community activity packages:
- UiPath.Database.Activities
- UiPath.FTP.Activities
- UiPath.Cryptography.Activities
- UiPath.Java.Activities
- UiPath.Credentials.Activities

coramiam committed:
Community Activities:

Added basic runtime telemetry (execution event only) to the following community activity packages:
- UiPath.Database.Activities
- UiPath.FTP.Activities
- UiPath.Cryptography.Activities
- UiPath.Java.Activities

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/DecryptFile.cs:
```diff
@@ -169,68 +179,75 @@ protected override void Execute(CodeActivityContext context)
 
-                    var item = new CryptographyLocalItem(decrypted, Path.GetFileName(outputFilePath), outputFilePath);
+                    // This overwrites the file if it already exists.
+                    File.WriteAllBytes(outputFilePath, decrypted);
```
` telemetryOperation?.Send();` should be right after this line

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/DecryptFile.cs:
```diff
@@ -169,68 +179,75 @@ protected override void Execute(CodeActivityContext context)
                 }
+                catch (Exception ex)
+                {
+                    Trace.TraceError(ex.ToString());
```
`telemetryOperation?.SendWithException(ex);` should be here

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/EncryptFile.cs:
```diff
@@ -169,57 +179,64 @@ protected override void Execute(CodeActivityContext context)
+                    }
+
+                    // This overwrites the file if it already exists.
+                    File.WriteAllBytes(outputFilePath, encrypted);
```
`telemetryOperation?.Send();` should be right after this line

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/EncryptFile.cs:
```diff
@@ -169,57 +179,64 @@ protected override void Execute(CodeActivityContext context)
                 }
+                catch (Exception ex)
+                {
+                    Trace.TraceError(ex.ToString());
```
`telemetryOperation?.SendWithException(ex);` should be here

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/EncryptText.cs:
```diff
@@ -140,25 +150,31 @@ protected override string Execute(CodeActivityContext context)
+                    var encrypted = CryptographyHelper.EncryptData(Algorithm, keyEncoding.GetBytes(input), CryptographyHelper.KeyEncoding(keyEncoding, key, keySecureString));
 
-                result = Convert.ToBase64String(encrypted);
+                    result = Convert.ToBase64String(encrypted);
```
`telemetryOperation?.Send();` should be right after this line

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/EncryptText.cs:
```diff
@@ -140,25 +150,31 @@ protected override string Execute(CodeActivityContext context)
+                }
+                catch (Exception ex)
+                {
+                    Trace.TraceError(ex.ToString());
```
`telemetryOperation?.SendWithException(ex);` should be here

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/HashFile.cs:
`telemetryOperation?.Send();` should be right after this line

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/HashFile.cs:
`telemetryOperation?.SendWithException(ex);` should be here

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/HashText.cs:
`telemetryOperation?.Send();` should be here

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/HashText.cs:
`telemetryOperation?.SendWithException(ex);` should be here

alexandru-petre commented on Activities/Cryptography/UiPath.Cryptography.Activities/UiPath.Cryptography.Activities.csproj:
```diff
@@ -6,10 +6,14 @@
 		<PackageReference Include="UiPath.Workflow" />
 		<PackageReference Include="System.Activities.Metadata" />
 		<PackageReference Include="System.Activities.ViewModels" />
+		<PackageReference Include="UiPath.Activities.Contracts" PrivateAssets="All" Version="24.10.1" />
```
Version should be in Directory.build.targets

```suggestion
		<PackageReference Include="UiPath.Activities.Contracts" PrivateAssets="All" />
```

alexandru-petre commented on Activities/Database/UiPath.Database.Activities/BulkInsert.cs:
```diff
@@ -37,56 +41,69 @@ public partial class BulkInsert : DatabaseRowActivity
+                }
+                catch (Exception ex)
+                {
+                    HandleException(ex, continueOnError);
```
`telemetryOperation?.SendWithException(ex);` should be here

alexandru-petre commented on Activities/Database/UiPath.Database.Activities/BulkInsert.cs:
```diff
@@ -37,56 +41,69 @@ public partial class BulkInsert : DatabaseRowActivity
+                    if (executorRuntime != null && executorRuntime.HasFeature(ExecutorFeatureKeys.LogMessage))
+                        return DbConnection.BulkInsertDataTable(tableName, dataTable, executorRuntime);
+                    else
+                        return DbConnection.BulkInsertDataTable(tableName, dataTable);
```
`telemetryOperation?.Send();` should be here

coramiam committed:
Merge branch 'feature/communityactivities/add-runtime-telemetry' of https://github.com/UiPath/Community.Activities into feature/communityactivities/add-runtime-telemetry

coramiam committed:
Addressed PR comments

Addressed the changes required for the PR
Fixed failing tests for Cryptography, Database, FTP and Java Activities

AlexMarinescuUiPath committed:
Refactor telemetry and session handling in WithFtpSession:
• Removed premature _telemetryOperation.Send() from ExecuteAsync to avoid duplicate success reporting (now only on completion).
• Assigned _ftpSession immediately after session creation (before OpenAsync) to ensure availability for fault cleanup.
• Adjusted scheduling to use existing _ftpSession without reassign inside delegate. No functional logic changed beyond telemetry flow and lifecycle handling.

AlexMarinescuUiPath committed:
added else branch for telemetry

AlexMarinescuUiPath committed:
reverted _ftpSession assignement to not introduce behavior change

AlexMarinescuUiPath committed:
fixed persistence of telemetryOperation

sorinconstantin commented on Activities/Shared/UiPath.Shared.Telemetry/ExecutionOperation.cs:
```diff
@@ -0,0 +1,62 @@
+            //    ActivityId = caller.Id,
+            //    WorkflowInstanceId = context.WorkflowInstanceId.ToString(),
+            //    WorkflowFilePath = context.GetExtension<IRunningJobInformation>()?.WorkflowFilePath ?? string.Empty,
+            //};
```
I would remove this commented out code

sorinconstantin commented on Activities/Shared/UiPath.Shared.Telemetry/Services/DesignTimeTelemetryService.cs:
```diff
@@ -0,0 +1,127 @@
+﻿//using System.Collections.Generic;
```
this should probably be removed too

sorinconstantin commented on Activities/Credentials/UiPath.Credentials.Activities/RequestCredential.cs:
```diff
@@ -37,29 +41,50 @@ public class RequestCredential : CodeActivity<bool>
 
-            var res = credPrompt.ShowDialog();
-            if (res != DialogResult.OK) return false;
+                telemetryOperation?.SetCustomDataKey(nameof(Message) + " IsUsed", message != null);
```
maybe we can put these strings in a constants class

AlexMarinescuUiPath replied:
made a constants class

sorinconstantin commented on Activities/Cryptography/UiPath.Cryptography.Activities.Tests/DecryptFileTests.cs:
```diff
@@ -3,6 +3,7 @@
 using System.Activities;
 using System.Security;
 using Xunit;
+using UiPath.Platform;
```
extra namespace

sorinconstantin commented on Activities/Cryptography/UiPath.Cryptography.Activities/DecryptText.cs:
```diff
@@ -140,34 +150,41 @@ protected override string Execute(CodeActivityContext context)
-                {
-                    throw;
-                }
+                telemetryOperation?.SendWithException(ex);
```
why is it necessary to send another Exception telemetry event?
it seems redundant

sorinconstantin commented on Activities/Cryptography/UiPath.Cryptography.Activities/EncryptFile.cs:
```diff
@@ -169,57 +179,64 @@ protected override void Execute(CodeActivityContext context)
-                Trace.TraceError(ex.ToString());
-
-                if (!ContinueOnError.Get(context)) throw;
+                telemetryOperation?.SendWithException(ex);
```
the same as above. Extra telemetry event for exception

sorinconstantin commented on Activities/Cryptography/UiPath.Cryptography.Activities/HashFile.cs:
```diff
@@ -54,33 +58,46 @@ protected override void CacheMetadata(CodeActivityMetadata metadata)
+                }
+            catch (Exception ex)
+            {
+                telemetryOperation?.SendWithException(ex);
```
The same pattern with an encapsulating extra try-catch block appears in many files

AlexMarinescuUiPath committed:
Refactor telemetry and exception handling

Added `UiPath.Shared.Telemetry` project to the solution and introduced `CredentialsConstants` for standardized telemetry keys. Refactored exception handling in cryptography activities to remove nested try-catch blocks and handle `ContinueOnError` logic. Removed unused `using` directive in `DecryptFileTests.cs` and redundant package references in project files. Added `UiPath.Telemetry.Client` to `UiPath.FTP.Activities.Packaging.csproj`. Removed `DesignTimeTelemetryService.cs` and related references. Streamlined code for improved maintainability and readability.

AlexMarinescuUiPath committed:
Refactor error handling and telemetry operations (#497)

Refactored multiple files to improve error handling and resource management by moving variable initialization outside of try blocks and using the HandleException method. Enhanced telemetry operations with custom data keys in DatabaseTransaction.cs. Removed redundant imports in UiPath.Credentials.Activities.csproj and added new shared project items in Activities.Database.sln.

sorinconstantin commented on Activities/FTP/UiPath.FTP.Activities/MoveItem.cs:
```diff
@@ -37,26 +38,40 @@ public partial class MoveItem : FtpCodeActivity
-                if (ftpSession == null)
+                PropertyDescriptor ftpSessionProperty = context.DataContext.GetProperties()[WithFtpSession.FtpSessionPropertyName];
+                IFtpSession ftpSession = ftpSessionProperty?.GetValue(context.DataContext) as IFtpSession;
+                try
```
one more "try" to remove

AlexMarinescuUiPath committed:
Add telemetry support to Python activities (#499)

Integrate `UiPath.Shared.Telemetry` service to enable telemetry
functionality in UiPath Python activities. Telemetry operations
are conditionally included using `#if ENABLE_DEFAULT_TELEMETRY`
directives. Enhance exception handling to capture and send
exceptions to the telemetry service before re-throwing. Update
`UiPath.Python.Activities.csproj` to import telemetry project
items, and modify the solution file to include telemetry
components. Ensure telemetry data is sent only when the
`ENABLE_DEFAULT_TELEMETRY` flag is defined.

AlexMarinescuUiPath committed:
Refactor telemetry handling and resource cleanup

Refactored telemetry operation handling in `ExecuteAsync`, `OnCompleted`, and `OnFaulted` methods to improve consistency and maintainability. Introduced conditional compilation (`#if ENABLE_DEFAULT_TELEMETRY`) for telemetry operations, reducing redundant code. Ensured proper sequencing of telemetry operations and resource cleanup (`_ftpSession.Close()` and `_ftpSession.Dispose()`) in all relevant methods.

AlexMarinescuUiPath committed:
Refactor and enhance database operation handling

Refactored code for improved readability, maintainability, and reliability. Standardized variable initialization, exception handling, and telemetry integration across methods. Consolidated database connection validation and ensured proper resource cleanup. Enhanced async execution patterns and parameter handling. Improved error logging and added validation for command timeouts. Updated bulk operations to support feature checks. Simplified and clarified code with additional comments.

AlexMarinescuUiPath committed:
Refactor Java activity execution and telemetry (#498)

* Refactor Java activity execution and telemetry

Refactored Java activity files to improve code structure by moving variable initializations outside of try blocks. This change affects `CreateJavaObject.cs`, `GetJavaField.cs`, `InvokeJavaMethod.cs`, and `LoadJar.cs`, ensuring that only method invocations are within try-catch blocks. Additionally, telemetry operations are now consistently sent after method invocations. Updated `JavaScope.cs` to handle telemetry operations in `OnFaulted` and `OnCompleted` methods. Added new shared project items to `Activities.Java.sln`.

* Refactor error handling and telemetry in async methods

Refactored `ExecuteAsync` methods across multiple files to improve error handling and ensure consistent telemetry operations. Moved initialization of key variables inside `try` blocks and wrapped method invocations in `try-catch` blocks to catch and log exceptions, throwing specific `InvalidOperationException` where applicable. Updated `JavaScope.cs` to send telemetry without preprocessor checks. These changes enhance code robustness and aid in monitoring and debugging.

* consistency update

AlexMarinescuUiPath committed:
Refactor and enhance error handling (#500)

* Refactor and enhance error handling
- Ensured consistency in exception handling and telemetry operations.

* Change for consistency

AlexMarinescuUiPath committed:
reverted telemetryOperation object creation

AlexMarinescuUiPath committed:
Merge pull request #502 from UiPath/feature/community/fix-database-telemetry

Refactor and enhance database operation handling

AlexMarinescuUiPath committed:
Merge pull request #501 from UiPath/feature/community/fix-ftp-telemetry

Refactor telemetry handling and resource cleanup

AlexMarinescuUiPath committed:
reverted telemetryObject initiatilzation to use context asap

AlexMarinescuUiPath committed:
possible build fix

AlexMarinescuUiPath committed:
simplified MoveItem

AlexMarinescuUiPath committed:
Add telemetry support to KeyedHashFile and KeyedHashText

AlexMarinescuUiPath committed:
Hash activities cleanup

AlexMarinescuUiPath committed:
Refactor: Remove legacy .NET Framework-specific code (#503)

* Refactor: Remove legacy .NET Framework-specific code

* Remove design-time components and localization files

This commit removes all design-time components and related files
from the `UiPath.Cryptography.Activities.Design` project.

AlexMarinescuUiPath committed:
build fix for cryptography

AlexMarinescuUiPath committed:
Add multi-targeting support for Windows frameworks

AlexMarinescuUiPath committed:
Simplify packaging by removing design-related files

AlexMarinescuUiPath committed:
reverted adding ;$(WindowsFramework)
